src/Processors/Formats/Impl/Parquet/ParquetBloomFilterCondition.cpp

```diff
@@ -0,0 +1,497 @@
+
+            rpn_stack.emplace_back(maybe_true, true);
+            if (element.function == Function::FUNCTION_NOT_IN)
+                rpn_stack.back() = !rpn_stack.back();
+        }
+        else if (element.function == Function::FUNCTION_NOT)
+        {
+            rpn_stack.back() = !rpn_stack.back();
+        }
+        else if (element.function == Function::FUNCTION_OR)
+        {
+            auto arg1 = rpn_stack.back();
+            rpn_stack.pop_back();
+            auto arg2 = rpn_stack.back();
+            rpn_stack.back() = arg1 | arg2;
+        }
+        else if (element.function == Function::FUNCTION_AND)
+        {
+            auto arg1 = rpn_stack.back();
+            rpn_stack.pop_back();
+            auto arg2 = rpn_stack.back();
+            rpn_stack.back() = arg1 & arg2;
+        }
+        else if (element.function == Function::ALWAYS_TRUE)
+        {
+            rpn_stack.emplace_back(true, false);
+        }
+        else
+        {
+            rpn_stack.emplace_back(true, true);
+        }
+    }
+
+    if (rpn_stack.size() != 1)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Unexpected stack size in KeyCondition::mayBeTrueOnRowGroup");
+
+    return rpn_stack[0].can_be_true;
+}
+
+std::unordered_set<std::size_t> ParquetBloomFilterCondition::getFilteringColumnKeys() const
+{
+    std::unordered_set<std::size_t> column_keys;
+
+    for (const auto & element : condition)
+    {
+        for (const auto index : element.key_columns)
+        {
+            column_keys.insert(index);
+        }
+    }
+
+    return column_keys;
+}
+
+/*
+ * `KeyCondition::rpn` is overly complex for bloom filters, some operations are not even supported. Not only that, but to avoid hashing each time
+ * we loop over a rpn element, we need to store hashes instead of where predicate values. To address this, we loop over `KeyCondition::rpn`
+ * and build a simplified RPN that holds hashes instead of values.
+ *
+ * `KeyCondition::RPNElement::FUNCTION_IN_RANGE` becomes:
+ *      `FUNCTION_EQUALS` when range limits are equal
+ *      `FUNCTION_UNKNOWN` when range limits are different
+ * `KeyCondition::RPNElement::FUNCTION_IN_SET` becomes
+ *      `FUNCTION_IN`
+ *
+ * Complex types and structs are not supported.
+ * There are two sources of data types being analyzed, and they need to be compatible: DB::Field type and parquet type.
+ * This is determined by the `isColumnSupported` method.
+ *
+ * Some interesting examples:
+ * 1. file(..., 'string_column UInt64') where str_column = '50'; Field.type == UInt64. Parquet type string. Not supported.
+ * 2. file(...) where str_column = 50; Field.type == String (conversion already taken care by `KeyCondition`). Parquet type string.
+ * 3. file(...) where uint32_column = toIPv4(5). Field.type == IPv4. Incompatible column types, resolved by `KeyCondition` itself.
+ * 4. file(...) where toIPv4(uint32_column) = toIPv4(5). Field.type == IPv4. We know it is safe to hash it using an int32 API.
```

Comment:
key_column_type is still uint32_t, regardless of the `toIPv4(uint32_column)`. Field.type has to match the type of the value it is being compared to, in this case, `toIPv4(...)`. Field.type is indeed `IPv4`.

what I meant to show here is that this comparison is still valid because we know the underlying bit representation of the data hasn't changed EVEN THOUGH it went through a transformation (i.e, a function call. `toIPv4`). Since the bit representation hasn't changed, the hash value won't change. Therefore, it is safe to allow this case.

This sounds like an exception, because most `toSomething` functions would change the bit representation and thus would be invalid. Note that I am talking about the lhs, the operation that takes place on the key_column.

Reply:
Oh, KeyCondition accepts this because toIPv4 is a monotonic function, and it puts the function in `RPNElement::monotonic_functions_chain`.

Wait, but then does it do the same for other functions? E.g. `negate(x) = -42` - `negate` is a monotonic function. I.e. RPNElement will have range = [-42, -42] and monotonic_functions_chain = {negate}. I checked, and this is indeed what happens. So keyConditionRPNToParquetBloomFilterCondition will incorrectly look for value -42 in the bloom filter instead of 42. You have to check if monotonic_functions_chain is empty and do FUNCTION_UNKNOWN if not. Please add a test to reproduce the error first.

(The same happens for `x + 10 = 52` - it handles binary function with one const argument.)

(Instead of FUNCTION_UNKNOWN we could try to apply the inverse function using IFunction::getPreimage, but that's not important.)